mono/utils/mono-coop-semaphore.c

```diff
@@ -0,0 +1,41 @@
+/**
+ * \file
+ */
+
+#include <config.h>
+#include "mono/utils/mono-coop-semaphore.h"
+#include "mono/utils/mono-threads.h"
+
+static void
+mono_coop_sem_interrupt (gpointer sem)
+{
+	mono_os_sem_post (&((MonoCoopSem*)sem)->s);
```

Comment:
This doesn't guarantee that it will wake up our waiter. It can wake up any other thread waiting on this semaphore and you can't really broadcast on a semaphore, it would also screw up the semaphore count. For these reasons, for example, I used a condvar in the monitor code instead of a semaphore.

Reply:
Interesting.
What if we undo 5f5c5e97a08f7086d7c18af37352c4a03dc4c0d1 but then
move the interruptibility into the coop_cond layer?

I was also debugging to see if there is was some other regression.
Like, @ludovic's hints were around the abort vs. suspend signal
and SA_RESTART and we should get back EINTR.

I see we do sometimes get EINTR, on Linux, but it doesn't work out for
reasons I don't yet know.

Mac is different. Signals are not used.